src/trace/spine_fueled.rs

```diff
@@ -0,0 +1,840 @@
+    // amortized work proportional to the size of batch.
+    fn insert(&mut self, batch: Self::Batch) {
+        assert!(batch.lower() != batch.upper());
+
+        // Ignore empty batches.
+        // Note: we may want to use empty batches to artificially force compaction.
+        if batch.is_empty() {
+            return;
+        }
+
+        self.dirty = true;
+
+        // Leonid: we do not require batch bounds to grow monotonically.
+        //assert_eq!(batch.lower(), &self.upper);
+
+        let index = batch.len().next_power_of_two();
+        self.introduce_batch(Some(batch), index.trailing_zeros() as usize);
+
+        // If more than one batch remains reschedule ourself.
+        if !self.reduced() {
+            if let Some(activator) = &self.activator {
+                activator.activate();
+            }
+        }
+    }
+
+    fn clear_dirty_flag(&mut self) {
+        self.dirty = false;
+    }
+
+    fn dirty(&self) -> bool {
+        self.dirty
+    }
+}
+
+impl<B> Spine<B>
+where
+    B: Batch + Clone + 'static,
+    B::Key: Ord,
+    B::Val: Ord,
+{
+    /// True iff there is at most one non-empty batch in `self.merging`.
+    ///
+    /// When true, there is no maintenance work to perform in the trace, other
+    /// than compaction. We do not yet have logic in place to determine if
+    /// compaction would improve a trace, so for now we are ignoring that.
+    fn reduced(&self) -> bool {
+        let mut non_empty = 0;
+        for index in 0..self.merging.len() {
+            if self.merging[index].is_double() {
+                return false;
+            }
+            if self.merging[index].len() > 0 {
+                non_empty += 1;
+            }
+            if non_empty > 1 {
+                return false;
+            }
+        }
+        true
+    }
+
+    /// Describes the merge progress of layers in the trace.
+    ///
+    /// Intended for diagnostics rather than public consumption.
+    #[allow(dead_code)]
+    fn describe(&self) -> Vec<(usize, usize)> {
+        self.merging
+            .iter()
+            .map(|b| match b {
+                MergeState::Vacant => (0, 0),
+                x @ MergeState::Single(_) => (1, x.len()),
+                x @ MergeState::Double(_) => (2, x.len()),
+            })
+            .collect()
+    }
+
+    /// Allocates a fueled `Spine` with a specified effort multiplier.
+    ///
+    /// This trace will merge batches progressively, with each inserted batch
+    /// applying a multiple of the batch's length in effort to each merge.
+    /// The `effort` parameter is that multiplier. This value should be at
+    /// least one for the merging to happen; a value of zero is not helpful.
+    pub fn with_effort(
+        mut effort: usize,
+        activator: Option<timely::scheduling::activate::Activator>,
+    ) -> Self {
+        // Zero effort is .. not smart.
+        if effort == 0 {
+            effort = 1;
+        }
```

Comment:
Maybe we should panic if given an effort of zero, that seems like a programmer error.
Alternatively we could take a `NonZeroUsize` instead of a `usize`, I like the type-safety route better